### ISSUE TITLE
Error reporter: Do not print source if source location spans whole file

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#8020]
The "missing pragma" error should not print anything from the source file, because it is irrelevant.

Component is SourceReferenceFormatter.

**Comment by yozaam** [comment]
Hello
I'm a smart contract developer and  newcomer to this codebase, but would love to make this contribution if I'm not wrong then this is something to do with the lines of : 
https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/libsolidity/analysis/SyntaxChecker.cpp#L71

 we can tell them to only display the message by making a change in error reporter?
 https://github.com/ethereum/solidity/blob/develop/liblangutil/ErrorReporter.cpp 

Sorry for being a newbie, but I would love to do this :D

**Comment by erak** [comment]
@yozaam First and foremost: thanks for being so eager making a contribution! The line you've identified is the point where the error gets reported initially. It is then later printed by the SourceReferenceFormatter or SourceReferenceFormatterHuman (see e.g. `SourceReferenceFormatter::printErrorInformation()`). I think at this point we need to check if the source location that should be printed spans over the entire file.

**Comment by yozaam** [comment]
> @yozaam First and foremost: thanks for being so eager making a contribution! The line you've identified is the point where the error gets reported initially. It is then later printed by the SourceReferenceFormatter or SourceReferenceFormatterHuman (see e.g. `SourceReferenceFormatter::printErrorInformation()`). I think at this point we need to check if the source location that should be printed spans over the entire file.

Thank you for this but is there a reason I should not do it from the liblangutil/ErrorReporter.cpp

Just check the location spans over entire file here and then not add it to errorList, Thanks a lot for helping me out 😄 
https://github.com/ethereum/solidity/blob/45caaf5ad8444d993389f56557763e4464a8c547/liblangutil/ErrorReporter.cpp#L62-L73

**Comment by erak** [comment]
@yozaam The issue with your proposal is, that it would prevent errors which span over the whole file being reported at all! As a result, no error would be reported in case the version pragma is missing. 

The fix need to be applied to the components that actually print those errors, as already stated in the initial description:

> The "missing pragma" error should not print anything from the source file, because it is irrelevant.
> 
> Component is SourceReferenceFormatter.

**Comment by yozaam** [comment]
> @yozaam The issue with your proposal is, that it would prevent errors which span over the whole file being reported at all! As a result, no error would be reported in case the version pragma is missing.
> 
> The fix need to be applied to the components that actually print those errors, as already stated in the initial description:
> 
> > The "missing pragma" error should not print anything from the source file, because it is irrelevant.
> > Component is SourceReferenceFormatter.

Ohh I get you thank you so much, I was thinking I would just send the description without location then they would only see the pragma error, okay so I have to check in SourceReferenceFormatter about the location

**Comment by yozaam** [comment]
https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/liblangutil/SourceReferenceFormatter.cpp#L87

This is the function, that sends the ref to the `printSourceLocation();`

And in printSourceLocation();
https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/liblangutil/SourceReferenceFormatter.cpp#L37-L43
Here we should put one more `else if(all lines of the program) { don't do anything }`


Also, can we see the .message contents and if it is from a pragma error, not call printSourceLocation here?
https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/liblangutil/SourceReferenceFormatter.cpp#L91

Wow, I hope I am not giving you too much trouble 😅

**Comment by erak** [comment]
@yozaam Besides `SourceReferenceFormatter`, the class `SourceReferenceFormatterHuman` probably needs to be touched as well. The latter one is used to generate human-readable error output, which is not necessarily given for `SourceReferenceFormatter`.

> https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/liblangutil/SourceReferenceFormatter.cpp#L87
> 
> This is the function, that sends the ref to the `printSourceLocation();`
> 
> And in printSourceLocation();
> 
> https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/liblangutil/SourceReferenceFormatter.cpp#L37-L43
> 
> Here we should put one more `else if(all lines of the program) { don't do anything }`
> 

I think applying the change you've sketched could be a good starting point.

> Also, can we see the .message contents and if it is from a pragma error, not call printSourceLocation here?
> 
> https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/liblangutil/SourceReferenceFormatter.cpp#L91

I don't think that this is needed. The error itself should be still printed even with your change. `printSourceLocation()` only handles (as the name says) source locations :wink:

You need to test your changes by adding a new folder to `/test/cmdlineTests/` and adding three files to it. An `input.sol` file, containing the source you want to test, an `err` file, containing the compiler output you expect (here without the source location), and an `exit` file, containing `1` as the expected return code. Please make sure to also check `/test/cmdlineTests/` for further details on how to write command line tests.

The above is needed to test the changes in `SourceReferenceFormatterHuman`, you can do all steps above for `SourceReferenceFormatter` by adding `--old-reporter` to the commandline arguments of the second test case in order to engage it instead of the new `SourceReferenceFormatterHuman`.

Tests can be run by invoking `./test/cmdlineTests.sh`. Please make sure all tests are passing.

> 
> Wow, I hope I am not giving you too much trouble sweat_smile

No worries, could you please open a pull request with your changes and the tests mentioned above? Thanks!

**Comment by yozaam** [comment]
I am very sorry, you are taking so much effort and I am not able to deliver, you could have solved this very easily but you were letting me learn, I spent all morning trying to understand how I could check if the source spans entire file
https://github.com/ethereum/solidity/blob/45caaf5ad8444d993389f56557763e4464a8c547/liblangutil/SourceReferenceExtractor.h#L41-L59
But there is only a check for multiline, and I am unable to figure out how to find out if it spans entire file 😞 
Running the test gives me 
```
y-./test/cmdlineTests.sh
Checking that the bug list is up to date...
Testing unknown options...
Incorrect response to unknown options: 
```

Since we have a MessageOnly function, are we to use that and only pass a nullptr as location for the error.warning when we get !m_pragmaVersionFound so it returns an empty location?

https://github.com/ethereum/solidity/blob/ba4a7244ed5707c8687d11e56fa9b395de32f130/liblangutil/SourceReferenceExtractor.cpp#L44-L50
Because of this, 
Call a empty location here?
https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/libsolidity/analysis/SyntaxChecker.cpp#L71
https://github.com/ethereum/solidity/blob/be0775da499415911b6be891e7ab60cc09e676c2/libsolidity/analysis/SyntaxChecker.cpp#L55-L71
 When I build it tells me  unused parameter '_sourceUnit'  so can i remove that from the function parameter?

Sorry Erik I am really trying my best

**Comment by a3d4** [comment]
A fix proposed in PR #8261.
